drivers/flash/flash_bflb.c

```diff
@@ -0,0 +1,1564 @@
+/* Register length of quad enable */
+	uint8_t  qe_write_reg_len;
+/* Register length of quad enable status */
+	uint8_t  qe_read_reg_len;
+/* Release power down command */
+	uint8_t  release_powerdown;
+/* Register length of contain busy status */
+	uint8_t  busy_read_reg_len;
+/* Read register command buffer */
+	uint8_t  read_reg_cmd[4];
+/* Write register command buffer */
+	uint8_t  write_reg_cmd[4];
+/* Enter qpi command */
+	uint8_t  enter_qpi;
+/* Exit qpi command */
+	uint8_t  exit_qpi;
+/* Config data for continuous read mode */
+	uint8_t  c_read_mode;
+/* Config data for exit continuous read mode */
+	uint8_t  c_rexit;
+/* Enable burst wrap command */
+	uint8_t  burst_wrap_cmd;
+/* Enable burst wrap command dummy clock */
+	uint8_t  burst_wrap_cmd_dmy_clk;
+/* Data and address mode for this command */
+	uint8_t  burst_wrap_data_mode;
+/* Data to enable burst wrap */
+	uint8_t  burst_wrap_data;
+/* Disable burst wrap command */
+	uint8_t  de_burst_wrap_cmd;
+/* Disable burst wrap command dummy clock */
+	uint8_t  de_burst_wrap_cmd_dmy_clk;
+/* Data and address mode for this command */
+	uint8_t  de_burst_wrap_data_mode;
+/* Data to disable burst wrap */
+	uint8_t  de_burst_wrap_data;
+/* Typical 4K(usually) erase time */
+	uint16_t time_e_sector;
+/* Typical 32K erase time */
+	uint16_t time_e_32k;
+/* Typical 64K erase time */
+	uint16_t time_e_64k;
+/* Typical Page program time */
+	uint16_t time_page_pgm;
+/* Typical Chip erase time in ms */
+	uint16_t time_ce;
+/* Release power down command delay time for wake up */
+	uint8_t  pd_delay;
+/* QE set data */
+	uint8_t  qe_data;
+} __packed;
+
+struct bflb_flash_header {
+	struct bflb_flash_magic_1 magic_1;
+	struct bflb_flash_magic_2 magic_2;
+	struct bflb_flash_cfg flash_cfg;
+	uint32_t flash_cfg_crc;
+} __packed;
+
+struct bflb_flash_command {
+/* Read write 0: read 1 : write */
+	uint8_t rw;
+/* Command mode 0: 1 line, 1: 4 lines */
+	uint8_t cmd_mode;
+/* SPI mode 0: IO 1: DO 2: QO 3: DIO 4: QIO */
+	uint8_t spi_mode;
+/* Address size */
+	uint8_t addr_size;
+/* Dummy clocks */
+	uint8_t dummy_clks;
+/* Transfer number of bytes */
+	uint32_t nb_data;
+/* Command buffer */
+	uint32_t cmd_buf[2];
+};
+
+struct flash_bflb_data {
+	struct bflb_flash_cfg flash_cfg;
+	uint32_t last_flash_offset;
+	uint32_t reg_copy;
+	uint32_t jedec_id;
+};
+
+/* Will using function cause error ? */
+static bool flash_bflb_is_in_xip(void *func)
+{
+	if ((uint32_t)func > BFLB_XIP_BASE && (uint32_t)func < BFLB_XIP_END) {
+		LOG_ERR("function at %d is in XIP and will crash the device", (uint32_t)func);
+		return true;
+	}
+
+	return false;
+}
+
+/* Are we doing something that makes sense? ? */
+static int flash_bflb_is_valid_range(off_t offset, size_t len)
+{
+	if (offset < 0) {
+		LOG_WRN("0x%lx: before start of flash", (long)offset);
+		return -EINVAL;
+	}
+	if ((offset + len) > TOTAL_SIZE) {
+		LOG_WRN("0x%lx: ends past the end of flash", (long)offset);
+		return -EINVAL;
+	}
```

Comment:
No need to separate checks, imho.
The check should be:
```
offset < 0 || len > TOTAL_SIZE || ((TOTAL_SIZE - offset) < len)
```
above avoids integer overflows.